### PR TITLE
MACE

### PR DESCRIPTION
**Pull request by RylieWeaver**
An implementation of the MACE Model in HYDRA.

There are 2 key aspects of the MACE Architecture:
(1) Message passing and interaction blocks are equivariant to the O(3) (rotation/inversion) group and invariant to the T(3) (translation) group.
(2) Predictions are made in an n-body expansion, where n is the number of layers. This is done by creating multi-body interactions, then decoding them. Layer 1 will decode 1-body interactions, layer 2 will decode w-body interactions, and so on. So, for a 3-layer model predicting energy, there are 3 outputs for energy, one at each layer, and they are summed at the end. This requires some adjustment to the behavior from Base.py. Specifically, there is a multihead decoder adapted from HYDRA after each convolutional layer.

### CODE REVIEW

File: .vscode/.vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
```
Comment (allaffa) [inline .vscode/.vscode/launch.json:1]: @RylieWeaver 
Is this file really needed or has it been added inadvertently?

File: .vscode/launch.json [hunk 2]
```diff
@@ -0,0 +1,74 @@
+{
```
Comment (allaffa) [inline .vscode/launch.json:1]: @RylieWeaver 
Is this file really needed or has it been added inadvertently?

File: examples/qm9/qm9.json [hunk 3]
```diff
@@ -5,12 +5,15 @@
     "NeuralNetwork": {
         "Profile": {"enable": 1},
         "Architecture": {
-            "model_type": "GIN",
+            "model_type": "MACE",
```
Comment (allaffa) [inline examples/qm9/qm9.json:8]: @RylieWeaver 
Have the changes to QM9 been added inadvertently?
Why are you changing existing examples?

**Review by allaffa — CHANGES_REQUESTED** [review]
Some files may have been added or changes by mistake. 
Please check and explain.

File: examples/qm9/qm9.json (hunk 3 above)
Comment (RylieWeaver) [reply examples/qm9/qm9.json:8]: @allaffa In all 3 of these files, they are inadvertent. Generally I clean these up later because if I fail the GitHub tests, then I still have the changes related to debugging.

If it's better, I can ping you when cleaned for review, but in general if you see something weird you are probably correct and it'll be removed later.

**Comment by RylieWeaver** [comment]
@allaffa As you can see, the MACE repo had a large collection of utils files. I'd like your opinion on how to handle these. On one hand, leaving too much can create clutter. On the other hand, taking things out could reduce customization options / induce unforeseen errors later on.

I think there is a middle-ground, where I could take out ~50% of the utils, focusing on dataset/distributed utils. In terms of impact, I estimate that it would only make a difference to a curious MACE user that uses an unusual argument, since it's hard to account for all these when cleaning out a big repo. In terms of time, it would take a day or two so still be ready by Monday to rebase.

**Comment by RylieWeaver** [comment]
@allaffa Additionally, despite having torch in their name, I have place `torch-ema` and `torchmetrics` in `requirements.txt`. I did this because the index-url for requirements-torch.txt was incompatible with thos packages. Is there a more preferable way to do this, or this is fine?

File: hydragnn/utils/mace_utils/data/hdf5_dataset.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    import h5py
+except ImportError:
+    pass
+
+from hydragnn.utils.mace_utils.data.atomic_data import AtomicData
+from hydragnn.utils.mace_utils.data.utils import Configuration
+from hydragnn.utils.mace_utils.tools.utils import AtomicNumberTable
+
+
+class HDF5Dataset(Dataset):
```
Comment (allaffa) [inline hydragnn/utils/mace_utils/data/hdf5_dataset.py:17]: @RylieWeaver 
here we will definitely need to interact with @jychoi-hpc to make sure that this dataset class interfaces well with the existing dataset infrastructure.

File: hydragnn/utils/mace_utils/data/neighborhood.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
+from typing import Optional, Tuple
```
Comment (allaffa) [inline hydragnn/utils/mace_utils/data/neighborhood.py:1]: @RylieWeaver 
What new capabilities does this add on top of the existing RadiusGraphPBC ?

File: hydragnn/utils/mace_utils/data/utils.py [hunk 6]
```diff
@@ -0,0 +1,398 @@
+###########################################################################################
```
Comment (allaffa) [inline hydragnn/utils/mace_utils/data/utils.py:1]: @RylieWeaver 
I see that these capabilities try to ensure compatibility between our implementation of MACE with the existing standalone open source implementation. 
However, is there any way to actually test the similarity between the two implementations? unless we have specific datasets to test this, it seems like this utils are somewhat dead-ended.

File: hydragnn/utils/mace_utils/modules/irreps_tools.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
+###########################################################################################
```
Comment (allaffa) [inline hydragnn/utils/mace_utils/modules/irreps_tools.py:1]: @RylieWeaver 
Do you think these utils are MACE-specific or can they be interchangeably used also for other equivariant models, namely NequIP?

File: hydragnn/utils/mace_utils/modules/models.py [hunk 8]
```diff
@@ -0,0 +1,1065 @@
+###########################################################################################
```
Comment (allaffa) [inline hydragnn/utils/mace_utils/modules/models.py:1]: @RylieWeaver 
This seems to be the native MACE implementation that we support as an isolated standalone set of scripts, correct? In this case, we could expand a bit further the comments at the top of the file to make this intention more explicit. For instance, we could explicitly mention the hyperlink to the files (mentioning also date and time we looked at those files, so that if those files change in the future, people know when to look back in the repository history). This could help especially in the perspective that the MACE developer may take a look at our code, as we did with theirs :)
It would be nice to come up with examples/scripts that can test the mathematical equivalence between this native implementation and our own integrated in the HydragNN framework.

File: hydragnn/utils/mace_utils/modules/models.py [hunk 9]
```diff
@@ -0,0 +1,1065 @@
...
+
+        output = {
+            "dipole": total_dipole,
+            "atomic_dipoles": atomic_dipoles,
+        }
+        return output
+
+
+@compile_mode("script")
+class EnergyDipolesMACE(torch.nn.Module):
```
Comment (allaffa) [inline hydragnn/utils/mace_utils/modules/models.py:819]: @RylieWeaver 
why is this model different from the regular MACE? why does it not inherit from the MACE class itself?

File: hydragnn/utils/mace_utils/modules/radial.py [hunk 10]
```diff
@@ -0,0 +1,323 @@
+###########################################################################################
```
Comment (allaffa) [inline hydragnn/utils/mace_utils/modules/radial.py:1]: @RylieWeaver 
Similarly to what asked for the `o3.irrep` functionalities, do these functionalities for Bessel function need to be MACE-specific or can they be made available also for other models. For instance, `DimeNet++` uses Bessel functions too [DmeNet++](https://github.com/ORNL/HydraGNN/blob/main/hydragnn/models/DIMEStack.py).
Can we make one implementation available for multiple models?

File: hydragnn/utils/mace_utils/modules/symmetric_contraction.py [hunk 11]
```diff
@@ -0,0 +1,233 @@
+###########################################################################################
```
Comment (allaffa) [inline hydragnn/utils/mace_utils/modules/symmetric_contraction.py:1]: @RylieWeaver 
Are also these functionalities MACE-specific or can they be made available for multiple models?

File: hydragnn/utils/mace_utils/tools/cg.py [hunk 12]
```diff
@@ -0,0 +1,131 @@
+###########################################################################################
```
Comment (allaffa) [inline hydragnn/utils/mace_utils/tools/cg.py:1]: @RylieWeaver 
Are also the functionalities for the Clebsch Gordan inner product MACE-specific or could they help support the implementation of other model?

File: hydragnn/utils/mace_utils/tools/scatter.py [hunk 13]
```diff
@@ -0,0 +1,112 @@
+"""basic scatter_sum operations from torch_scatter from
```
Comment (allaffa) [inline hydragnn/utils/mace_utils/tools/scatter.py:1]: @RylieWeaver 
Are these implementations different from what the package `torch-scatter` natively provides?
Please take a look at this [torch-scatter](https://github.com/rusty1s/pytorch_scatter/blob/master/csrc/scatter.cpp)

File: hydragnn/utils/mace_utils/tools/torch_geometric/batch.py [hunk 14]
```diff
@@ -0,0 +1,257 @@
+from collections.abc import Sequence
```
Comment (allaffa) [inline hydragnn/utils/mace_utils/tools/torch_geometric/batch.py:1]: @RylieWeaver 
Do we need this as standalone routine instead of using the native implementation provided by `torch-geometric`?

**Review by allaffa — CHANGES_REQUESTED** [review]
Second round of review completed.

File: hydragnn/utils/mace_utils/tools/torch_geometric/batch.py (hunk 14 above)
Comment (RylieWeaver) [reply hydragnn/utils/mace_utils/tools/torch_geometric/batch.py:1]: @allaffa Want to meet on Teams to discuss?

File: hydragnn/utils/config_utils.py [hunk 15]
```diff
@@ -109,23 +127,23 @@ def update_config(config, train_loader, val_loader, test_loader):
...
+        ), "E(3) equivariance can only be ensured for EGNN, SchNet, and MACE."
     elif "equivariance" not in config:
         config["equivariance"] = False
     return config
 
 
 def update_config_edge_dim(config):
     config["edge_dim"] = None
-    edge_models = ["PNAPlus", "PNA", "CGCNN", "SchNet", "EGNN"]
+    edge_models = ["PNAPlus", "PNA", "CGCNN", "SchNet", "EGNN", "MACE"]
```
Comment (allaffa) [inline hydragnn/utils/config_utils.py:142]: @RylieWeaver 
Could we add also `DimeNet++` to this list?

File: hydragnn/utils/config_utils.py [hunk 16]
```diff
@@ -109,23 +127,23 @@ def update_config(config, train_loader, val_loader, test_loader):
...
     if "edge_features" in config and config["edge_features"]:
         assert (
             config["model_type"] in edge_models
-        ), "Edge features can only be used with EGNN, SchNet, PNA, PNAPlus, and CGCNN."
+        ), "Edge features can only be used with MACE, EGNN, SchNet, PNA, PNAPlus, and CGCNN."
```
Comment (allaffa) [inline hydragnn/utils/config_utils.py:146]: @RylieWeaver 
Could we add also `DimeNet++` to this list?

**Review by allaffa — COMMENTED** [review]
Just a coupled of questions about why `DimeNet++` is missing from some tests

File: hydragnn/utils/config_utils.py (hunk 15 above)
Comment (RylieWeaver) [reply hydragnn/utils/config_utils.py:146]: Hmm, double-checking that DImeNet supports edge feats...

Comment (RylieWeaver) [reply hydragnn/utils/config_utils.py:146]: @allaffa DimeNet's architecture doesn't natively support edge_feats. It would be possible to customize the `InteractionPPBlock` to do it (~1 days work) if it's of interest. However, it would not be the original architecture.

**Comment by RylieWeaver** [comment]
@allaffa 

There is something important to note wrt `blocks.py` in MACE:

Many of these blocks need to be changed in order to adapt to HYDRA (usually shaping and variable passing).  Importantly, many of these "blocks" are similar, with small tweaks. For example:
```
    RealAgnosticAttResidualInteractionBlock,
    RealAgnosticInteractionBlock,
    RealAgnosticResidualInteractionBlock,
    ResidualElementDependentInteractionBlock,
```

The differences between these blocks are a residual connection or not, and similarly small adjustments. Given the similarities and in order to make it more tractable, I chose to adjust and implement one of these blocks, which I discern to be generally applicable and effective. For example, I chose `RealAgnosticAttResidualInteractionBlock` out of those 4.

TLDR: This should be generally effective to choose one block and move forward, however is important to note down to memory in case an impassioned MACE user comes around with a very specific choice of architecture.

**Comment by allaffa** [comment]
> @allaffa
> 
> There is something important to note wrt `blocks.py` in MACE:
> 
> Many of these blocks need to be changed in order to adapt to HYDRA (usually shaping and variable passing). Importantly, many of these "blocks" are similar, with small tweaks. For example:
> 
> ```
>     RealAgnosticAttResidualInteractionBlock,
>     RealAgnosticInteractionBlock,
>     RealAgnosticResidualInteractionBlock,
>     ResidualElementDependentInteractionBlock,
> ```
> 
> The differences between these blocks are a residual connection or not, and similarly small adjustments. Given the similarities and in order to make it more tractable, I chose to adjust and implement one of these blocks, which I discern to be generally applicable and effective. For example, I chose `RealAgnosticAttResidualInteractionBlock` out of those 4.
> 
> TLDR: This should be generally effective to choose one block and move forward, however is important to note down to memory in case an impassioned MACE user comes around with a very specific choice of architecture.

@RylieWeaver 
I agree with your choice and thank you for letting me know.

**Comment by RylieWeaver** [comment]
@allaffa Here are the main updates on MACE:

Progress: 
(1) It is rebased to the recent utils PR
(2) Almost all the low-hanging fruit to delete has been taken out

Next Steps:
(1) Investigate unnecessary or broadly applicable functions. For example, `scatter_add` might be importable from torch_scatter, and `cg.py` might have things used in Nequip.
(2) Fix `data.x` use. Although the tests are passing, the one-hot encoding I apply could cause errors in general cases. This will be changed, although it's worth noting it's **highly recommended** that node features like atom type are one-hot-encoded when input to MACE. I'll put this in a comment and `RaiseWarning` somewhere

Notes:
(1) I've changed a lot of the test/json files just so it runs faster, but I'll revert these before the final draft.
(2) Re: It's highly recommended that node_features like atom type are one-hot encoded when input into MACE

File: hydragnn/utils/model/mace_utils/modules/irreps_tools.py [hunk 17]
```diff
@@ -0,0 +1,86 @@
+###########################################################################################
```
Comment (allaffa) [inline hydragnn/utils/model/mace_utils/modules/irreps_tools.py:1]: @RylieWeaver 
Specify from which source you got this file

File: hydragnn/utils/model/mace_utils/modules/blocks.py [hunk 18]
```diff
@@ -0,0 +1,443 @@
+###########################################################################################
```
Comment (allaffa) [inline hydragnn/utils/model/mace_utils/modules/blocks.py:1]: @RylieWeaver 
Specify form which source you got this file.

File: hydragnn/utils/model/mace_utils/modules/radial.py [hunk 19]
```diff
@@ -0,0 +1,323 @@
+###########################################################################################
```
Comment (allaffa) [inline hydragnn/utils/model/mace_utils/modules/radial.py:1]: @RylieWeaver 
Specify from which source you got this file.

File: hydragnn/utils/model/mace_utils/modules/symmetric_contraction.py [hunk 20]
```diff
@@ -0,0 +1,233 @@
+###########################################################################################
```
Comment (allaffa) [inline hydragnn/utils/model/mace_utils/modules/symmetric_contraction.py:1]: @RylieWeaver 
Specify from which source you got this file.

File: hydragnn/utils/model/mace_utils/modules/utils.py [hunk 21]
```diff
@@ -0,0 +1,48 @@
+###########################################################################################
```
Comment (allaffa) [inline hydragnn/utils/model/mace_utils/modules/utils.py:1]: @RylieWeaver 
Specify from which source you got this file.

File: hydragnn/utils/model/mace_utils/tools/cg.py [hunk 22]
```diff
@@ -0,0 +1,131 @@
+###########################################################################################
```
Comment (allaffa) [inline hydragnn/utils/model/mace_utils/tools/cg.py:1]: @RylieWeaver 
Specify from which source you got this file.

File: hydragnn/utils/model/mace_utils/tools/scatter.py [hunk 23]
```diff
@@ -0,0 +1,112 @@
+"""basic scatter_sum operations from torch_scatter from
```
Comment (allaffa) [inline hydragnn/utils/model/mace_utils/tools/scatter.py:1]: @RylieWeaver 
Remember to check if these functionalities are already provided in the new versions of the package `torch-scatter `

File: tests/test_graphs.py [hunk 24]
```diff
@@ -196,16 +197,17 @@ def unittest_train_model(
...
-        "GIN",
-        "GAT",
-        "MFC",
-        "PNA",
-        "PNAPlus",
-        "CGCNN",
-        "SchNet",
-        "DimeNet",
-        "EGNN",
+        # "SAGE",
```
Comment (allaffa) [inline tests/test_graphs.py:200]: @RylieWeaver 
You commented all the other models out. I guess this is one for now since the pR is in draft mode. However, remember to uncomment them as soon as the PR needs to be ready for review.

File: tests/test_graphs.py [hunk 25]
```diff
@@ -214,26 +216,49 @@ def pytest_train_model(model_type, ci_input, overwrite_data=False):
 
 
 # Test only models
-@pytest.mark.parametrize("model_type", ["PNA", "PNAPlus", "CGCNN", "SchNet", "EGNN"])
+@pytest.mark.parametrize(
+    # "model_type", ["PNA", "PNAPlus", "CGCNN", "SchNet", "EGNN", "MACE"]
```
Comment (allaffa) [inline tests/test_graphs.py:220]: @RylieWeaver 
You commented all the other models out. I guess this is one for now since the pR is in draft mode. However, remember to uncomment them as soon as the PR needs to be ready for review.

File: tests/test_graphs.py [hunk 26]
```diff
@@ -214,26 +216,49 @@ def pytest_train_model(model_type, ci_input, overwrite_data=False):
...
     unittest_train_model(model_type, "ci_vectoroutput.json", True, overwrite_data)
 
 
-@pytest.mark.parametrize(
-    "model_type",
-    ["SAGE", "GIN", "GAT", "MFC", "PNA", "PNAPlus", "SchNet", "DimeNet", "EGNN"],
-)
-def pytest_train_model_conv_head(model_type, overwrite_data=False):
-    unittest_train_model(model_type, "ci_conv_head.json", False, overwrite_data)
+# @pytest.mark.parametrize(
```
Comment (allaffa) [inline tests/test_graphs.py:242]: @RylieWeaver 
You commented all the other models out. I guess this is one for now since the pR is in draft mode. However, remember to uncomment them as soon as the PR needs to be ready for review.

**Review by allaffa — COMMENTED** [review]
A few additional comments left.

File: hydragnn/utils/model/mace_utils/modules/utils.py (hunk 21 above)
Comment (RylieWeaver) [reply hydragnn/utils/model/mace_utils/modules/utils.py:1]: Should it have more than this?:

```
###########################################################################################
# Utilities
# Authors: Ilyes Batatia, Gregor Simm and David Kovacs
# This program is distributed under the MIT License (see MIT.md)
###########################################################################################
```

Comment (RylieWeaver) [reply hydragnn/utils/model/mace_utils/modules/utils.py:1]: Do you just mean to link it?

File: hydragnn/utils/mace_utils/modules/models.py (hunk 9 above)
Comment (RylieWeaver) [reply hydragnn/utils/mace_utils/modules/models.py]: MACE has a bunch of models that are slightly different for different applications. I took this one out and others, and kept the general-case MACE.

File: hydragnn/utils/mace_utils/data/hdf5_dataset.py (hunk 4 above)
Comment (RylieWeaver) [reply hydragnn/utils/mace_utils/data/hdf5_dataset.py]: Unnecessary deletion

File: hydragnn/utils/mace_utils/data/neighborhood.py (hunk 5 above)
Comment (RylieWeaver) [reply hydragnn/utils/mace_utils/data/neighborhood.py]: Unnecessary Deletion

File: hydragnn/utils/mace_utils/data/utils.py (hunk 6 above)
Comment (RylieWeaver) [reply hydragnn/utils/mace_utils/data/utils.py]: Unnecessary Deletion

File: hydragnn/utils/mace_utils/tools/torch_geometric/batch.py (hunk 14 above)
Comment (RylieWeaver) [reply hydragnn/utils/mace_utils/tools/torch_geometric/batch.py]: Unnecessary Deletion

File: hydragnn/utils/model/mace_utils/tools/cg.py (hunk 22 above)
Comment (allaffa) [reply hydragnn/utils/model/mace_utils/tools/cg.py:1]: @RylieWeaver 
In the last meeting, we had discussed the idea of including the hyperlink in the description long with the date on which the file has been taken from the original repo. This way, if the file is changed later, people can revert to the version that we took ourselves.

Comment (RylieWeaver) [reply hydragnn/utils/model/mace_utils/tools/cg.py:1]: @allaffa Ah, I see now. I had this in `MACESTACK.py` interpreted this as sufficient, but can add this to the other files.

File: hydragnn/utils/model/mace_utils/modules/__init__.py [hunk 27]
```diff
@@ -0,0 +1,59 @@
+###########################################################################################
+# __init__ file for Modules
+# Authors: Ilyes Batatia, Gregor Simm
+# This program is distributed under the MIT License (see MIT.md)
+###########################################################################################
+# Taken From:
```
Comment (allaffa) [inline hydragnn/utils/model/mace_utils/modules/__init__.py:6]: @RylieWeaver 
Yes, I like this. Please apply to all the other files as well.

**Review by allaffa — COMMENTED** [review]
Getting closer. :)

**Comment by RylieWeaver** [comment]
@allaffa 

> Getting closer. :)

Yep :). Here’s a summary update for posterity, which I’ll also cover in the meeting:

---

1. Replacing `torch_scatter` operations with an import worked fine. The original implementation avoided it for dependency reasons, but we’re good as long as we can install `torch_scatter`.

2. I moved `irreps_tools.py` to the `utils/model` directory for broader applicability and created `operations.py` for a general function to get `edge_vectors_and_lengths`. We might revisit the naming or location.

3. I couldn’t find anything similar in Nequip to generalize further. For example, `grep` searches for "wigner," "U_matrix," and "Contraction" returned nothing. Nequip's `InteractionBlock` is also different.

4. In `compile.py`, I left all compilation-related functions to ensure speed and accuracy, even though some can be commented out without issue.

**Comment by RylieWeaver** [comment]
@allaffa At this point, I have made all changes which I can think of. So, I think the steps left are:

(1) Review by you, making any changes requested
(2) Revert the changes I made to json and test_***.py files for quicker testing
(3) Last review by you and Pei
(4) Merge

**Comment by RylieWeaver** [comment]
@allaffa 

There is an important update to note here which I've realized. It is that **_MACE only natively supports atomic_numbers as node_attributes_**. I am following up with the author, but I believe it is unfeasible or non-trivial or to change this.

At the current version that I have pushed now, it is a requirement that only raw atomic numbers are passed as node_attributes. I do the one-hot encoding automatically.

I also include tests in `process_node_attributes` function of `MACEStack.py` which will warn or throw an error if the data looks off.

File: hydragnn/utils/model/mace_utils/modules/blocks.py [hunk 28]
```diff
@@ -0,0 +1,394 @@
...
+        )
+
+    def __repr__(self):
+        return (
+            f'{self.__class__.__name__}(shape=({", ".join(str(s) for s in self.weights.shape)}), '
+            f"weights={np.prod(self.weights.shape)})"
+        )
+
+
+@compile_mode("script")
```
Comment (allaffa) [inline hydragnn/utils/model/mace_utils/modules/blocks.py:275]: @RylieWeaver 
Add a comment line above this function to explain that the original MACE had also different alternative interactions blocks. We should explain that this is only one of the many interaction blocks available, and that the other blocks can be retrieved from the original MACE implementation.

File: hydragnn/utils/model/mace_utils/modules/radial.py [hunk 29]
```diff
@@ -0,0 +1,331 @@
...
+            )
+        else:
+            self.register_buffer("gaussian_weights", gaussian_weights)
+        self.coeff = -0.5 / (r_max / (num_basis - 1)) ** 2
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:  # [..., 1]
+        x = x - self.gaussian_weights
+        return torch.exp(self.coeff * torch.pow(x, 2))
+
+
```
Comment (allaffa) [inline hydragnn/utils/model/mace_utils/modules/radial.py:116]: @RylieWeaver 
Is this applied to avoid abrupt cutoff of the interactions and therefore ensure differentiability? 
If my understanding is right, please add a few sentences on top to of this function to explain this.

File: hydragnn/utils/model/mace_utils/modules/radial.py [hunk 30]
```diff
@@ -0,0 +1,331 @@
...
+        # noinspection PyUnresolvedReferences
+        return envelope * (x < self.r_max)
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}(p={self.p}, r_max={self.r_max})"
+
+
+@compile_mode("script")
+class ZBLBasis(torch.nn.Module):
```
Comment (allaffa) [inline hydragnn/utils/model/mace_utils/modules/radial.py:151]: @RylieWeaver 
is this used currently? Do you foresee the need to use it in the future? if yes, why?

File: hydragnn/utils/model/operations.py [hunk 31]
```diff
@@ -0,0 +1,35 @@
+##############################################################################
```
Comment (allaffa) [inline hydragnn/utils/model/operations.py:1]: @RylieWeaver 
What do you think about including this function into [hydragnn/preprocess/graph_samples_checks_and_updates.py](https://github.com/ORNL/HydraGNN/blob/main/hydragnn/preprocess/graph_samples_checks_and_updates.py)?

**Review by allaffa — COMMENTED** [review]
One more round of review completed.

File: hydragnn/utils/model/mace_utils/modules/radial.py (hunk 29 above)
Comment (RylieWeaver) [reply hydragnn/utils/model/mace_utils/modules/radial.py:116]: @allaffa 

No, that is done outside of the specific basis type. Specifically, the RadialEmbeddingBlock has these lines:

```
        if radial_type == "bessel":
            self.bessel_fn = BesselBasis(r_max=r_max, num_basis=num_bessel)
        elif radial_type == "gaussian":
            self.bessel_fn = GaussianBasis(r_max=r_max, num_basis=num_bessel)
        elif radial_type == "chebyshev":
            self.bessel_fn = ChebychevBasis(r_max=r_max, num_basis=num_bessel)
        if distance_transform == "Agnesi":
            self.distance_transform = AgnesiTransform()
        elif distance_transform == "Soft":
            self.distance_transform = SoftTransform()
        self.cutoff_fn = PolynomialCutoff(r_max=r_max, p=num_polynomial_cutoff)
```

```
return radial * cutoff
```

So, the cutoff is applied outside of the exact type of radial function. These specific lines perform the Gaussian Smearing for radial basis function, where the `self.gaussian_weights` are the gaussian centers.

File: hydragnn/utils/model/mace_utils/modules/radial.py [hunk 32]
```diff
@@ -0,0 +1,331 @@
...
+            v_edges, receiver, dim=0, dim_size=node_attrs.size(0), reduce="sum"
+        )
+        return V_ZBL.squeeze(-1)
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}(r_max={self.r_max}, c={self.c})"
+
+
+@compile_mode("script")
+class AgnesiTransform(torch.nn.Module):
```
[the remark this answers is not in the document]
Comment (RylieWeaver) [reply hydragnn/utils/model/mace_utils/modules/radial.py:234]: @allaffa 

Not regularly that I have seen. Similar to the ZBLBasis, it serves as just another hyperparameter for the way to transform and use distance.

File: hydragnn/utils/model/operations.py (hunk 31 above)
Comment (RylieWeaver) [reply hydragnn/utils/model/operations.py:1]: @allaffa 

Hmm, I am unsure. If we move it into the `preprocess` directory, would we want to execute it during the preprocess steps, rather than during `_conv_args` function? That would make strict sense, but would be slightly more complicated. How do you think it would look in this way?

File: requirements-torch2.txt [hunk 33]
```diff
@@ -0,0 +1,3 @@
+e3nn
```
[the remark this answers is not in the document]
Comment (RylieWeaver) [reply requirements-torch2.txt]: @allaffa

Ok, I have changed it so that they are all in `requirement-torch.txt`. This requires 2 things:

(1) An additional url to be passed to the pip install in `CI.yml`. This is where the pip install looks if it can't find the package in the first url.
(2) The specific package version to be specified. This way, an outdated package does not get found and installed from the first url.

**Comment by RylieWeaver** [comment]
Ready for next round of review

File: hydragnn/utils/config_utils.py (hunk 15 above)
Comment (allaffa) [reply hydragnn/utils/config_utils.py]: @RylieWeaver 
Would it be ok to include this option in DimeNet++? Obviously this would be addressed in a separate PR, not this one.

**Comment by allaffa** [comment]
> @allaffa
> 
> > Getting closer. :)
> 
> Yep :). Here’s a summary update for posterity, which I’ll also cover in the meeting:
> 
> 1. Replacing `torch_scatter` operations with an import worked fine. The original implementation avoided it for dependency reasons, but we’re good as long as we can install `torch_scatter`.
> 2. I moved `irreps_tools.py` to the `utils/model` directory for broader applicability and created `operations.py` for a general function to get `edge_vectors_and_lengths`. We might revisit the naming or location.
> 3. I couldn’t find anything similar in Nequip to generalize further. For example, `grep` searches for "wigner," "U_matrix," and "Contraction" returned nothing. Nequip's `InteractionBlock` is also different.
> 4. In `compile.py`, I left all compilation-related functions to ensure speed and accuracy, even though some can be commented out without issue.

@RylieWeaver 
Since  the functionalities in`compile.py` can be commented out without issue and are provided only for the sake of speed-up, I think it would be beneficial to add a short description about this at the beginning of the file.

**Review by allaffa — COMMENTED** [review]
I left a few more comments.

File: hydragnn/utils/config_utils.py (hunk 15 above)
Comment (RylieWeaver) [reply hydragnn/utils/config_utils.py]: @allaffa 
So, adjust DimeNet to allow for edge_feats? Yes, I could do this and put it on my to-do. Just like the message to confirm.

File: hydragnn/utils/model/operations.py (hunk 31 above)
Comment (allaffa) [reply hydragnn/utils/model/operations.py:1]: I see what you mean. Let us keep it as is.

File: hydragnn/utils/model/mace_utils/modules/radial.py (hunk 32 above)
[the remark this answers is not in the document]
Comment (allaffa) [reply hydragnn/utils/model/mace_utils/modules/radial.py:234]: @RylieWeaver 
If we keep it as an alternatively option, then we should create a test which is regularly run by the CI where these bases are used. This is the only way to check that these functionalities will not be broken by future changes in the repository. 
Keeping these things hanging around without actually using them is very risky and prone to having things around that will easily become deprecated and not supported anymore.

**Review by allaffa — COMMENTED** [review]
Another round completed.

File: hydragnn/utils/model/mace_utils/modules/radial.py (hunk 30 above)
Comment (allaffa) [reply hydragnn/utils/model/mace_utils/modules/radial.py:151]: @RylieWeaver 
I would add a comment to the file saying that this basis is useful for modeling extremely close interactions, where nuclear forces dominate ([https://arxiv.org/abs/1904.00360]).

Also, I think we should have a dedicated test in the CI where we continuously check that the MACE model can switch between different choice of basis functions, so that we make sure that these basis are maintained in our code and their use does not become deprecated.

Comment (RylieWeaver) [reply hydragnn/utils/model/mace_utils/modules/radial.py:151]: @allaffa 

(1) After further looking into it, I believe I was partially mistaken on the ZBL basis. It is useful for modeling close interactions, but is for predicting atomic energies, not serving as a distance transformation. Given that Hydra is application-agnostic and the niche use-case, I believe it should just be removed.

(2) I made a new test file `test_radial_transforms.py` to test all the different possible transformations. In the future, other models with multiple possible transformations could be put in here as well. Let me know if this is what you're thinking, or if another way is better.

File: tests/test_graphs.py [hunk 34]
```diff
@@ -196,16 +197,17 @@ def unittest_train_model(
...
-        "GAT",
-        "MFC",
-        "PNA",
-        "PNAPlus",
-        "CGCNN",
-        "SchNet",
-        "DimeNet",
-        "EGNN",
+        # "SAGE",
+        # "GIN",
```
Comment (allaffa) [inline tests/test_graphs.py:201]: @RylieWeaver 
You MUST absolutely get rid of this comments once for all. 
 "temporarily" commenting things out like this is extremely dangerous.

File: tests/test_graphs.py [hunk 35]
```diff
@@ -214,26 +216,49 @@ def pytest_train_model(model_type, ci_input, overwrite_data=False):
...
 
 
-@pytest.mark.parametrize(
-    "model_type",
-    ["SAGE", "GIN", "GAT", "MFC", "PNA", "PNAPlus", "SchNet", "DimeNet", "EGNN"],
-)
-def pytest_train_model_conv_head(model_type, overwrite_data=False):
-    unittest_train_model(model_type, "ci_conv_head.json", False, overwrite_data)
+# @pytest.mark.parametrize(
+#     "model_type",
```
Comment (allaffa) [inline tests/test_graphs.py:243]: @RylieWeaver 
Uncomment what needs to be kept, and remove everything else.

**Review by allaffa — COMMENTED** [review]
Another round of review

**Comment by RylieWeaver** [comment]
@allaffa 

Ready for next round. All tests should be running now.

**Comment by allaffa** [comment]
@RylieWeaver 
Conflicts left to resolve

**Comment by RylieWeaver** [comment]
> @RylieWeaver Conflicts left to resolve

Oh, it's from the typo fix just now. The struggle of multiple PRs, lol

File: tests/test_radial_transforms.py [hunk 36]
```diff
@@ -0,0 +1,208 @@
...
+#                                                                            #
+# SPDX-License-Identifier: BSD-3-Clause                                      #
+##############################################################################
+
+import sys, os, json
+import pytest
+
+import torch
+
+torch.manual_seed(97)
```
Comment (allaffa) [inline tests/test_radial_transforms.py:17]: @RylieWeaver 
Is there any reason why you chose 97 ?

**Review by allaffa — APPROVED** [review]
I left just a question about the torch seed set to 97. 
For the rest, it looks good to me.

File: tests/test_radial_transforms.py (hunk 36 above)
Comment (RylieWeaver) [reply tests/test_radial_transforms.py:17]: @allaffa 

Nope, this is just an artifact of me using the code from `test_graphs.py` to make this test.

**Comment by RylieWeaver** [comment]
> I left just a question about the torch seed set to 97. For the rest, it looks good to me.

Shall we have Pei take a look for a final check with some fresh eyes?

**Review by ArCho48 — APPROVED** [review]
